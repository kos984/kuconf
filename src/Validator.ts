import * as assert from 'assert';
import * as _ from 'lodash';
import * as Validator from 'validatorjs';

export enum ECastType {
  Number = 'number',
  Boolean = 'boolean',
  Date = 'Date',
  String = 'string',
}

export enum ERule {
  Required = 'required',
  Date = 'date',
  Numeric = 'numeric',
  String = 'string',
  Min = 'min',
  Max = 'max',
  Integer = 'integer',
  /** value should be strict equals 'true' or 'false' */
  Bool = 'kuConfBool',
  /**
   * object should have properties only in listed values;
   * objectKeysIn:Y,N will check attrs: {a: 'Y', b: 'N'}
   */
  ObjectKeysIn = 'kuConfObjectKeysIn',
  /**
   * cast:ECastType,defaultValue
   *
   * @example
   * const rules = { '*.age': `numeric|${ERule.Cast}:${ECastType.Number},10` };
   * const data = [{ age: '4' }, { name: 'test' }, { age: 5 }];
   *
   * @description should be last rule in the list, will not update inputValue
   */
  Cast = 'kuConfCast',
}

Validator.register(
  ERule.Bool,
  (value: any) => {
    return typeof value === 'boolean' || value === 'true' || value === 'false';
  },
  'The :attribute must be a true or false',
);

Validator.register(
  ERule.ObjectKeysIn,
  function(value: { [key: string]: string }, params: string, attribute: string) {
    const ruleIn = this.validator.getRule('in');
    for (const key of Object.keys(value)) {
      const val = value[key];
      const rulePassed = ruleIn.validate(val, params, `${attribute}.${key}`);
      if (!rulePassed) {
        return false;
      }
    }
    return true;
  },
  'The :attribute keys should be in :objectKeysIn',
);

/** should be last rule in a queue, will not update inputValue */
(Validator as any).registerImplicit(
  ERule.Cast,
  function(value: any, input: any, path: string) {
    const source = this.validator.input;
    const [type, defaultValue] = input.split(',');
    if (!_.has(source, path)) {
      if (defaultValue) {
        value = defaultValue;
      } else {
        return true;
      }
    } else if (typeof value !== 'string') {
      // cast string only
      return true;
    }
    switch (type) {
      case ECastType.Number: {
        value = typeof value === 'number' ? value : Number(value);
        break;
      }
      case ECastType.Boolean: {
        value = typeof value === 'boolean' ? value : value === 'true';
        break;
      }
      case ECastType.Date: {
        value = new Date(value);
        break;
      }
      case ECastType.String: {
        value = value.toString();
        break;
      }
      default: {
        return false;
      }
    }
    _.set(source, path, value);
    return true;
  },
  'Cast rule not supports',
);

// test cast
{
  const data = [{ age: '4' }, { name: 'test' }, { age: 5 }];
  const rules = { '*.age': `numeric|${ERule.Cast}:${ECastType.Number}` };
  const validation = new Validator(data, rules);
  validation.passes(); // true
  assert.deepEqual(data, [{ age: 4 }, { name: 'test' }, { age: 5 }], 'Validator, cast rule is not works correctly');
}

export default Validator;
