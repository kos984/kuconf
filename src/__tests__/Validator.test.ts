import Validator, { ECastType, ERule } from '../Validator';

describe('Validator', () => {

  it(`rule: ${ERule.Bool}`, () => {
    const validator = new Validator({
      arr: [ 'true', 'false', true, false, 'True', 'False', 'any' ],
    }, {
      'arr.*': ERule.Bool,
    });
    validator.fails();
    expect(validator.errors).toMatchSnapshot();
  });

  it(`rule: ${ERule.ObjectKeysIn}`, () => {
    const validator = new Validator({
      arr: [
        { prop: 'Y' },
        { prop: 'N' },
        { prop: 'A' },
        { prop: 'B' },
      ],
    }, {
      'arr.*': `${ERule.ObjectKeysIn}:Y,N`,
    });
    validator.fails();
    expect(validator.errors).toMatchSnapshot();
  });

  it(`rule: ${ERule.Cast}`, () => {
    const payload: any = {
      bool: 'true',
      date: '1999-12-31',
      nullDate: null,
      number: '100',
    };
    const validator = new Validator(payload, {
      // should cast if value exists
      bool: `${ERule.Bool}|${ERule.Cast}:${ECastType.Boolean}`,
      date: `${ERule.Date}|${ERule.Cast}:${ECastType.Date}`,
      nullDate: `${ERule.Cast}:${ECastType.Date}`, // should be null
      number: `${ERule.Numeric}|${ERule.Cast}:${ECastType.Number}`,

      // should set default property if value not exists
      boolDefaultFalse: `${ERule.Bool}|${ERule.Cast}:${ECastType.Boolean},false`,
      boolDefaultTrue: `${ERule.Bool}|${ERule.Cast}:${ECastType.Boolean},true`,
      dateDefault: `${ERule.Date}|${ERule.Cast}:${ECastType.Date}:1999-12-31`,
      dateTimeDefault: `${ERule.Cast}:${ECastType.Date},1999-12-31T00:00:00`,
      numberDefault: `${ERule.Numeric}|${ERule.Cast}:${ECastType.Number},100`,
      stringDefault: `${ERule.String}|${ERule.Cast}:${ECastType.String},default\,String`,

      // should not set if property not exists
      boolEmpty: `${ERule.Bool}|${ERule.Cast}:${ECastType.Boolean}`,
      dateEmpty: `${ERule.Date}|${ERule.Cast}:${ECastType.Date}`,
      numberEmpty: `${ERule.Numeric}|${ERule.Cast}:${ECastType.Number}`,
    });
    validator.fails();
    expect(validator.errors).toMatchSnapshot();
    expect(payload).toMatchSnapshot();
  });

});
