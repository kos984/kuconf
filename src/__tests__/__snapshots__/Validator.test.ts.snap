// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator rule: bool 1`] = `
Errors {
  "errors": Object {
    "arr.4": Array [
      "The arr.4 must be a true or false",
    ],
    "arr.5": Array [
      "The arr.5 must be a true or false",
    ],
    "arr.6": Array [
      "The arr.6 must be a true or false",
    ],
  },
}
`;

exports[`Validator rule: cast 1`] = `
Errors {
  "errors": Object {},
}
`;

exports[`Validator rule: cast 2`] = `
Object {
  "bool": true,
  "boolDefaultFalse": false,
  "boolDefaultTrue": true,
  "date": 1999-12-31T00:00:00.000Z,
  "dateTimeDefault": 1999-12-31T00:00:00.000Z,
  "number": 100,
  "numberDefault": 100,
  "stringDefault": "default",
}
`;

exports[`Validator rule: objectKeysIn 1`] = `
Errors {
  "errors": Object {
    "arr.2": Array [
      "The arr.2 keys should be in Y,N",
    ],
    "arr.3": Array [
      "The arr.3 keys should be in Y,N",
    ],
  },
}
`;
